packages/angular_devkit/build_angular/src/builders/web-test-runner/index.ts

```diff
@@ -0,0 +1,199 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import { BuilderContext, BuilderOutput, createBuilder } from '@angular-devkit/architect';
+// eslint-disable-next-line import/no-duplicates
+import type * as WebTestRunner from '@web/test-runner';
+// eslint-disable-next-line import/no-duplicates
+import type { EventEmitter } from '@web/test-runner';
+import { promises as fs } from 'fs';
+import path from 'path';
+import { findTestFiles } from '../../utils/test-files';
+import { buildApplicationInternal } from '../application';
+import { OutputHashing } from '../browser-esbuild/schema';
+import { logBuilderStatusWarnings } from './builder-status-warnings';
+import { WtrBuilderOptions, normalizeOptions } from './options';
+import { Schema } from './schema';
+
+export default createBuilder(
+  async (schema: Schema, ctx: BuilderContext): Promise<BuilderOutput> => {
+    ctx.logger.warn(
+      'NOTE: The Web Test Runner builder is currently EXPERIMENTAL and not ready for production use.',
+    );
+    logBuilderStatusWarnings(schema, ctx);
+
+    const options = normalizeOptions(schema);
+    const testDir = 'dist/test-out';
```

Comment:
This is okay for now, but we should move away from writting spec files to disk and inside wire up WRT to read the files from in-memory similar to how we handle Karma and Vite.

Reply:
Agreed, I'll need to look into whether and how to pass through a virtual file system to Web Test Runner.